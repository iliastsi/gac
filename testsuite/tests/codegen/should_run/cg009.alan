solve () : proc

   hanoi (rings : int, src : reference byte [],
          trg : reference byte [], aux : reference byte []) : int

      move (src : reference byte [], trg : reference byte []) : proc
      {
         writeString("moving from ");
         writeString(src);
         writeString(" to ");
         writeString(trg);
	 writeString("\n");
      }

      n1 : int;
      n2 : int;

   { -- hanoi
      if (rings > 0) {
         n1 = hanoi(rings-1, src, aux, trg);
         move(src, trg);
         n2 = hanoi(rings-1, aux, trg, src);
         return n1 + 1 + n2;
      }
      else
         return 0;
   } -- hanoi

   rings : int;
   moves : int;

{ -- solve
   writeString("Please, give the number of rings: ");
   rings = readInteger();
   moves = hanoi(rings, "left", "right", "middle");
   writeString("Solved in ");
   writeInteger(moves);
   writeString(" move(s).\n");
} -- solve
