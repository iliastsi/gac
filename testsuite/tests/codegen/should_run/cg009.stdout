Please, give the number of rings: moving from left to middle
moving from left to right
moving from middle to right
moving from left to middle
moving from right to left
moving from right to middle
moving from left to middle
moving from left to right
moving from middle to right
moving from middle to left
moving from right to left
moving from middle to right
moving from left to middle
moving from left to right
moving from middle to right
Solved in 15 move(s).
